proof/invariant-abstract/SchedContext_AI.thy

```diff
@@ -555,20 +558,89 @@ lemma sc_consumed_update_sym_refs[wp]:
    \<lbrace>\<lambda>_ s. P (state_refs_of s)\<rbrace>"
   by (wpsimp wp: update_sched_context_refs_of_same)
 
-lemma refill_unblock_check_valid_state [wp]:
-  "\<lbrace>valid_state\<rbrace> refill_unblock_check r \<lbrace>\<lambda>rv. valid_state\<rbrace>"
-  apply (wpsimp simp: refill_unblock_check_def get_refills_def if_apply_def2 wp: hoare_drop_imps hoare_vcg_all_lift)
-  apply (clarsimp simp: valid_state_def refills_merge_valid[simplified])
+crunches refill_unblock_check
+  for ex_nonz_cap_to[wp]: "\<lambda>s. ex_nonz_cap_to ptr s"
+  (simp: crunch_simps is_round_robin_def wp: crunch_wps)
+
+lemma refill_unblock_check_if_live_then_nonz_cap[wp]:
+  "refill_unblock_check usage \<lbrace>if_live_then_nonz_cap\<rbrace>"
+  apply (clarsimp simp: refill_unblock_check_defs)
+  apply (wpsimp wp: whileLoop_wp' get_refills_wp update_sched_context_wp hoare_drop_imps)
   done
 
-lemma refill_unblock_check_cur_tcb [wp]:
-  "\<lbrace>cur_tcb\<rbrace> refill_unblock_check r \<lbrace>\<lambda>rv. cur_tcb\<rbrace>"
-  by (wpsimp simp: refill_unblock_check_def if_apply_def2 wp: hoare_drop_imps hoare_vcg_all_lift)
+crunches refill_unblock_check
```

Comment:
At first glance it looks like there's a lot of crossover between this big `crunches` and the earlier one for `refill_budget_check`. It might be a long shot, but any chance of combining them?

Reply:
Nope :-(